pkg/bitmap/bitmap.go

```diff
@@ -0,0 +1,232 @@
+
+// Package bitmap provides the implementation of bitmap.
+package bitmap
+
+import (
+	"math"
+	"math/bits"
+)
+
+// +stateify savable
+type Bitmap struct {
+	numOnes   int
+	bitBlock []uint64
+}
+
+// NewBitmap create a new empty Bitmap.
+func NewBitmap(size int) *Bitmap {
+	b := new(Bitmap)
+	bSize := size / 64
+	if size % 64 != 0 {
+		bSize = bSize + 1
+	}
+	b.bitBlock = make([]uint64, bSize, bSize)
+	return b
+}
+
+// IsEmpty verifiys if the Bitmap is empty.
+func (b *Bitmap) IsEmpty() bool {
+	return b.numOnes == 0
+}
+
+// Minimum return the smallest value in the Bitmap.
+func (b *Bitmap) Minimum() int32 {
+	for i := 0; i < len(b.bitBlock); i++ {
+		if w := b.bitBlock[i]; w != 0 {
+			r := bits.TrailingZeros64(w)
+			return int32(r + i*64)
+		}
+	}
+	return math.MaxInt32
+}
+
+// Maximum return the largest value in the Bitmap.
+func (b *Bitmap) Maximum() int32 {
+	for i := len(b.bitBlock); i > 0; i-- {
+		if w := b.bitBlock[i-1]; w != 0 {
+			r := bits.LeadingZeros64(w)
+			return int32((i-1)*64 + 63 - r)
+		}
+	}
+	return int32(0)
+}
+
+// Add add i to the Bitmap.
+func (b *Bitmap) Add(i int32) {
+	blockNum, mask := i/64, uint64(1)<<(i%64)
+	// if blockNum is out of range, extend b.bitBlock
+	if x, y := int(blockNum), len(b.bitBlock); x >= y {
+		b.bitBlock = append(b.bitBlock, make([]uint64, x-y+1)...)
+	}
+	oldBlock := b.bitBlock[blockNum]
+	newBlock := oldBlock | mask
+	if oldBlock != newBlock {
+		b.bitBlock[blockNum] = newBlock
+		b.numOnes++
+	}
+}
+
+// Remove i from the Bitmap.
+func (b *Bitmap) Remove(i int32) {
+	blockNum, mask := i/64, uint64(1)<<(i%64)
+	oldBlock := b.bitBlock[blockNum]
+	newBlock := oldBlock &^ mask
+	if oldBlock != newBlock {
+		b.bitBlock[blockNum] = newBlock
+		b.numOnes--
+	}
+}
+
+// Clone the Bitmap.
+func (b *Bitmap) Clone() *Bitmap {
+	ptr := Bitmap{b.numOnes, make([]uint64, len(b.bitBlock))}
+	copy(ptr.bitBlock, b.bitBlock[:])
+	return &ptr
+}
+
+// ReplaceWith replace the content of b with that of newBitmap.
+func (b *Bitmap) ReplaceWith(newBitmap *Bitmap) {
+	b.numOnes = newBitmap.numOnes
+	b.bitBlock = make([]uint64, len(newBitmap.bitBlock))
+	copy(b.bitBlock, newBitmap.bitBlock[:])
+}
+
+// countOnesForBlocks count all 1 bits within b.bitBlock of begin and that of end.
+// The begin block and end block are inclusive.
+func (b *Bitmap) countOnesForBlocks(begin, end int) uint64 {
+	ones := uint64(0)
+	if begin > end {
+		return ones
+	}
+	beginBlock := begin / 64
+	endBlock := end / 64
+	for i := beginBlock; i <= endBlock; i++ {
+		ones += uint64(bits.OnesCount64(b.bitBlock[i]))
+	}
+	return ones
+}
+
+// countOnesForAllBlocks count all 1 bits in b.bitBlock.
+func (b *Bitmap) countOnesForAllBlocks() uint64 {
+	ones := uint64(0)
+	for i := 0; i < len(b.bitBlock); i++ {
+		ones += uint64(bits.OnesCount64(b.bitBlock[i]))
+	}
+	return ones
+}
+
+// flipBitmapRange flip the bits within range (begin and end). begin and end are inclusive.
+func (b *Bitmap) flipBitmapRange(begin, end int) {
```

Comment:
nit: Now since this method is **on** the BitMap, we can drop the "BitMap" from the function name. Rename it to just `flipRange`. Same for `removeBitmapRange`.